### ISSUE TITLE
Upgrading causes GLB to become invalid

### ISSUE THREAD

**Issue by javagl** [CesiumGS/3d-tiles-validator#213]
This is the result of diving into a rabbit hole of debugging, and I'll try to provide some context, but keep it short:

There are some "legacy features" in 3D Tiles that are still used by other tools and existing assets, for example, the `gltfUpAxis` property in the `asset`. The fact that Cesium for Unreal does not honor this property caused some issues with rotated models, at https://community.cesium.com/t/unreal-load-3dtileset-not-by-ion-looks-wrong-rotation/12835 .  The recommendation was to upgrade the tileset with `3d-tiles-validator`. The upgrade seems to work, but the resulting B3DMs contain GLBs that are invalid (and are not rendered by Cesium for Unreal and other tools - some viewers still manage to render them, though)

To reproduce: (Note: Example output files will be given below)

- Unpack the data set that is linked in the forum thread
- Delete the `tileset.json` files in all subdirectories (otherwise, the conversion will complain with "More than one root tileset found in directory" - that should probably be fixed as well...)
- Upgrade the data set with
  `node ./bin/3d-tiles-tools.js upgrade --input C:/Data/3dtiles/ --output C:/Data/3dtiles_Upgraded`
- Extract GLBs from one B3DM from the **OLD** and the **NEW** data set with
  `node ./bin/3d-tiles-tools.js b3dmToGlb --input C:\Data\3dtiles\Tile_+000_+001\Tile_+000_+001_L20_00000t3.b3dm --output C:\Data\Tile_+000_+001_L20_00000t3-OLD.glb`
  and 
 `node ./bin/3d-tiles-tools.js b3dmToGlb --input C:\Data\3dtiles_Upgraded\Tile_+000_+001\Tile_+000_+001_L20_00000t3.b3dm --output C:\Data\Tile_+000_+001_L20_00000t3-NEW.glb`
- Drag the resulting GLBs into the glTF validator at https://github.khronos.org/glTF-Validator/ 

For the `OLD` version, it prints an unrelated error about the buffer view target. For the `NEW` one, the validator reports 

            {
                "code": "GLB_CHUNK_TOO_BIG",
                "message": "Chunk (0x004e4942) length (906568) does not fit total GLB length.",
                "severity": 0,
                "offset": 1528
            },
            {
                "code": "GLB_LENGTH_MISMATCH",
                "message": "Declared length (908100) does not match GLB length (908104).",
                "severity": 0,
                "offset": 908104
            },
            {
                "code": "BUFFER_GLB_CHUNK_TOO_BIG",
                "message": "GLB-stored BIN chunk contains 4 extra padding byte(s).",
                "severity": 1,
                "pointer": "/buffers/0"
            },

(omitted unrelated warnings here). 

Since the data set was shared publicly in the forum, I hope that it is OK to attach the relevant files here. (`OLD` and `NEW` referring to *before* and *after* the upgrade).

[3d-tiles-validator-chunk-size-issue-data.zip](https://github.com/CesiumGS/3d-tiles-validator/files/6250886/3d-tiles-validator-chunk-size-issue-data.zip)

**Comment by baothientran** [comment]
Offline discussion about the problem

> @javagl The issue of assets containing the gltfUpVector (causing a wrong rotation) has some ripple effects. (Even though the gltfUpVector was only part of the pre-1.0 3D Tiles spec for a short time, when something is "public", it can never be taken back (at least, not without causing a lot of pain)).
Upgrading the tileset with 3d-tiles-validator partially works, but has some issues, resulting in invalid GLB files. These are rightfully rejected by our implementation. There seems to be only 4 unnecessary padding bytes being inserted, and brutally changing https://github.com/CesiumGS/cesium-native/blob/main/CesiumGltfReader/src/Reader.cpp#L278 to
```

    if (binaryEnd > glbData.size()) {
      // XXX
      binaryChunk = glbData.subspan(binaryStart, glbData.size() - binaryStart);
    /*
      return {
          std::nullopt,
          {"GLB binary chunk extends past the end of the buffer, binary end "
           "at " +
           std::to_string(binaryEnd) + ", data size " +
           std::to_string(glbData.size())},
          {}};
      */
    }
    else 
    {
      binaryChunk = glbData.subspan(binaryStart, pBinaryChunkHeader->chunkLength);
    }

```

> to just ignore these bytes causes the model to be rendered, but that's not a viable path. There also doesn't seem to be a roadmap for fixes or consolidation in gltf-pipeline or 3d-tiles-validator.
> Considering the trade-off, I'd vote for adding support of the gltfUpVector property in Cesium for Unreal. It's legacy. It's not official. But it's hard to ignore that it exists.


> @baothientran I think it is because after padding the glb buffer, it doesn't fix the glb header length https://github.com/CesiumGS/3d-tiles-validator/blob/56fa6278369d393f2fc8d9bff14f51400c116c0b/tools/lib/getGlbPadded.js#L26

> it fixes the chunk, but the header is left out

**Comment by baothientran** [comment]
so after fixing the glb header length, it seems to render on unreal. The only problem is the gltf's buffer is now mismatched with the chunk

**Comment by javagl** [comment]
OK, it took remarkably long for me to understand what you guys have been talking about. A summary:

- The GLB *inside* the B3DM must start and end on 8-byte boundaries (according to https://github.com/CesiumGS/3d-tiles/tree/master/specification/TileFormats/Batched3DModel#padding). 
- In the [`createB3dm` function](https://github.com/CesiumGS/3d-tiles-validator/blob/56fa6278369d393f2fc8d9bff14f51400c116c0b/tools/lib/createB3dm.js#L23), this is achieved by padding the first buffer (after the header) to end at an 8-byte-boundary, and padding all remaining buffers so that their length is a multiple of 8. 
- The `getGlbPadded` adds the necessary padding bytes at the *end* of the GLB, but does so by making the `BIN` chunk of the GLB longer.

And referring to the layout image from the spec at https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#binary-gltf-layout : The function only updated the `chunkLength` of the binary buffer, but it did **not** update the `length` from the main (12-byte) GLB header. 

@baothientran I'm not sure whether I fully understood that part

> The only problem is the gltf's buffer is now mismatched with the chunk

If this is only about the fact that the `BIN` chunk is now **longer** than the actual `buffer.byteLength`, then this is OK - quoting from https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#glb-stored-buffer :

> byte length of `BIN` chunk could be up to 3 bytes bigger than JSON-defined `buffer.byteLength` to satisfy GLB padding requirements.

(I could now try to wrap my head around the question of whether it could be up to **7** bytes for our case, because of the 8-byte-alignment requirement from B3DM, and whether this would be spec-compliant, but ... maybe later). 

If there is a fix that makes sure that the GLBs are valid, then this could also help to fix the issue with the other data set that is not rendered properly right now. I have not yet figured out why it is **not** rendered in its original form, but after running the "upgrade" script (and when ignoring the GLB issue), it is rendered. Further details have to be investigated.

**Comment by baothientran** [comment]
> (I could now try to wrap my head around the question of whether it could be up to 7 bytes for our case, because of the 8-byte-alignment requirement from B3DM, and whether this would be spec-compliant, but ... maybe later).

@javagl yup, but in our case, the chunk is up-to 4 bytes bigger than JSON-defined `buffer.byteLength`. It's because GLB chunk requires to be aligned at 4 bytes boundary at the begin and end. So If we add a correct GLB buffer to b3dm, there will be 4 bytes missing if GLB is padded to be aligned at 8 bytes boundary
